### PR TITLE
Parsing improvements to Cisco Parser

### PR DESCRIPTION
**Pull request by dhalperi**
At the current stage, every new rule has new configs to be tested. (In the same commit as each parser change).

### CODE REVIEW

File: projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/ConfigurationFormat.java [hunk 1]
```diff
@@ -8,6 +8,7 @@
   ARISTA("arista"),
   AWS_VPC("aws_vpc"),
   BLADENETWORK("bladenetwork"),
+  CADANT("cadant"),
```
Comment (arifogel) [inline projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/ConfigurationFormat.java]: Also requires cases in `org.batfish.datamodel.RoutingProtolL:getDefaultAdministrativeCost` and `getSummaryAdministrativeCost`

File: projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/CiscoParser.g4 [hunk 2]
```diff
@@ -1438,6 +1438,16 @@ s_flow
    )*
 ;
 
+s_flow_sampler_map
+:
+   NO? FLOW_SAMPLER_MAP ~NEWLINE* NEWLINE s_fsm_mode?
```
Comment (arifogel) [inline projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/CiscoParser.g4]: Change `s_fsm_mode` to `fsm_mode`

File: projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/CiscoParser.g4 [hunk 3]
```diff
@@ -1236,6 +1236,7 @@ s_archive
          | LOG
```
Comment (arifogel) [inline projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/CiscoParser.g4]: This innermost parenthetical expression should be moved to a new rule `a_null` or something similar if that is already taken.

File: projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/CiscoParser.g4 [hunk 4]
```diff
@@ -1334,6 +1338,7 @@ s_dial_peer
 
 s_domain
```
Comment (arifogel) [inline projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/CiscoParser.g4]: Needs an `enterS_domain` function in `CiscoControlPlaneExtractor` setting `_no` to `ctx.NO() != null`. The `exitS_domain` function should set `_no` to `false`.
Anything `enter.*` or `exit.*`functions for parser rules under `s_domain` need to check value of `_no` before performing action.

File: projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_line.g4 [hunk 5]
```diff
@@ -32,6 +32,13 @@ l_exec_timeout
    EXEC_TIMEOUT minutes = DEC seconds = DEC? NEWLINE
 ;
 
+l_length
```
Comment (arifogel) [inline projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_line.g4]: Broken, and must be resolved against #388

File: projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_common.g4 [hunk 6]
```diff
@@ -293,6 +293,7 @@ port
    | KSHELL
    | LDAP
    | LDAPS
+   | LDP
```
Comment (arifogel) [inline projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_common.g4]: Need to add `else if` case to `org.batfish.grammar.cisco.CiscoControlPlaneExtractor:toNamedPort`. Then you may need to add data model `NamedPort`.

File: projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_routemap.g4 [hunk 7]
```diff
@@ -68,6 +68,11 @@ boolean_as_path_in_rp_stanza
    AS_PATH IN expr = as_path_set_expr
 ;
 
+boolean_as_path_is_local_rp_stanza
```
Comment (arifogel) [inline projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_routemap.g4]: Needs handling in `CiscoControlPlaneExtractor and all the way through pipeline.
```java
private RoutePolicyBoolean toRoutePolicyBoolean(Boolean_simple_rp_stanzaContext ctx)
```

File: projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/CiscoParser.g4 (hunk 2 above)
Comment (dhalperi) [reply projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/CiscoParser.g4]: done

File: projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_common.g4 (hunk 6 above)
Comment (dhalperi) [reply projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_common.g4]: done

File: projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_line.g4 (hunk 5 above)
Comment (dhalperi) [reply projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_line.g4]: fixed

File: projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_routemap.g4 (hunk 7 above)
Comment (dhalperi) [reply projects/batfish/src/main/antlr4/org/batfish/grammar/cisco/Cisco_routemap.g4]: done

**Comment by dhalperi** [comment]
Added parsing examples for all new commands, updated parsing ref and data model diffs.

**Pull request merged**